documenting external logins (#327)
front-commerce/developers.front-commerce.com: source/docs/advanced/features/external-logins.md

---
id: external-logins
title: External Logins
---

<blockquote class="feature--new">
_Since version 2.13_
</blockquote>

<blockquote class="feature--new">

> _Beta feature (api may change)_

Reviewer: ```suggestion
_Beta feature (API may change)_
```

</blockquote>

# External Logins

Front Commerce supports external logins. The API for supporting external logins comes in 2 parts:
* The `LoginProvider`s which implements the login with the external systems (such as Facebook or Google)
* The `ExternalLoginHandler` which implement the login with your platform (such as Magento1, Magento2, Proximis or BigCommerce).

> Out of the box we ship two `LoginProvider`s -the `FacebookProvider` and the `GoogleProvider`- and two `ExternalLoginHandler`s -the `Magento1ExternalLoginHandler` and the `Magento2ExternalLoginHandler`.

Reviewer: isn't it a bit too detailed for something that may/will change ?

Author: i think what would change will be more on the implementaion lavel the concepts of LoginProvider and LoginHandler will remain

After ensuring that `AdditionalLoginFormActions` and `FlashMessages` are in your `LoginForm` the "Sign in with Facebook" and "Sign in with Google" buttons should now appear in your login page/modal.

## Adding the stylesheet

For the out of the box login buttons you can include the following stylesheet in your `main.scss` file:

```diff
...
@import "~theme/modules/modules";
@import "~theme/pages/pages";
+@import "~theme/external-login.scss";
...
```

## Configuring the buttons

To configure the buttons shown for external logins you can override the `theme/modules/User/LoginForm/AdditionalLoginFormActions/AdditionalLoginFormActions.js` file and add or remove the buttons you want.

## Configuring the Facebook provider

The `FacebookProvider` needs the `clientId` and the `clientSecret` to work. You can get these from your app on the [facebook developer console](https://developers.facebook.com/apps/).

To configure the `clientId` and the `clientSecret` you need to define the folowing environment variables:
* `FRONT_COMMERCE_FACEBOOK_CLIENT_ID` for the `clientId`
* `FRONT_COMMERCE_FACEBOOK_CLIENT_SECRET` for the `clientSecret`

### Configuring Redirect URIs

> You need to configure the allowed redirect URIs on your [facebook developer console](https://developers.facebook.com/apps/). To do so go you need to select your app and then go to `Facebook Login > Settings` section on the left menu. There you will find a field called `Valid OAuth Redirect URIs`. You need to fill this field with your stores URLs folowed by `/external-login/facebook/callback`. For example if you have 2 stores `https://example.com/en` and `https://example.com/fr`, you need to put `https://example.com/en/external-login/facebook/callback` and `https://example.com/fr/external-login/facebook/callback` in the `Valid OAuth Redirect URIs` field.

Author: the nuance here is:
* needing to register all the stores
* postfix the stores urls by `/external-login/facebook/callback`

these they cant find in facebook docs

> You need to configure the allowed redirect URIs on your [facebook developer console](https://developers.facebook.com/apps/). To do so go you need to select your app and then go to `Facebook Login > Settings` section on the left menu. There you will find a field called `Valid OAuth Redirect URIs`. You need to fill this field with your stores URLs folowed by `/external-login/facebook/callback`. For example if you have 2 stores `https://example.com/en` and `https://example.com/fr`, you need to put `https://example.com/en/external-login/facebook/callback` and `https://example.com/fr/external-login/facebook/callback` in the `Valid OAuth Redirect URIs` field.

Reviewer 2: Maybe this would be clearer then ?

```suggestion
You need to configure the allowed redirect URIs on your [facebook developer console](https://developers.facebook.com/apps/). 

Note that the `Valid OAuth Redirect URIs` should contain all your stores URLs followed by `/external-login/facebook/callback`. 
 
For example for stores `https://example.com/en` and `https://example.com/fr`, you need to put `https://example.com/en/external-login/facebook/callback` and `https://example.com/fr/external-login/facebook/callback` in the `Valid OAuth Redirect URIs` field.
```

> You need to configure the allowed redirect URIs on your [facebook developer console](https://developers.facebook.com/apps/). To do so go you need to select your app and then go to `Facebook Login > Settings` section on the left menu. There you will find a field called `Valid OAuth Redirect URIs`. You need to fill this field with your stores URLs folowed by `/external-login/facebook/callback`. For example if you have 2 stores `https://example.com/en` and `https://example.com/fr`, you need to put `https://example.com/en/external-login/facebook/callback` and `https://example.com/fr/external-login/facebook/callback` in the `Valid OAuth Redirect URIs` field.

Author: that works thx

> You need to configure the allowed redirect URIs on your [facebook developer console](https://developers.facebook.com/apps/). To do so go you need to select your app and then go to `Facebook Login > Settings` section on the left menu. There you will find a field called `Valid OAuth Redirect URIs`. You need to fill this field with your stores URLs folowed by `/external-login/facebook/callback`. For example if you have 2 stores `https://example.com/en` and `https://example.com/fr`, you need to put `https://example.com/en/external-login/facebook/callback` and `https://example.com/fr/external-login/facebook/callback` in the `Valid OAuth Redirect URIs` field.

Author: also updated the google one in the same spirit

_Note: do not forget to [reconfigure your app](#Configuring-Redirect-URIs) with this new URL_

After starting the proxying service and configuring your app and `FRONT_COMMERCE_FACEBOOK_DEVELOPMENT_BASE_CALLBACK_URL` accordingly, the "Login with Facebook" button should work properly.

## Configuring the Google provider

The `GoogleProvider` needs the `clientId` and the `clientSecret` to work. You can get these from your "OAuth2.0 Client" of your app on the [google developer console](https://console.developers.google.com/apis/credentials). P.S. you need to create one if you have not already.

To configure the `clientId` and the `clientSecret` you need to define the folowing environment variables:
* `FRONT_COMMERCE_GOOGLE_CLIENT_ID` for the `clientId`
* `FRONT_COMMERCE_GOOGLE_CLIENT_SECRET` for the `clientSecret`

### Configuring Redirect URIs

> You need to configure the allowed redirect URIs under your "OAuth2.0 Client" of your app on the [google developer console](https://console.developers.google.com/apis/credentials). Fill the `Authorized redirect URIs` field with your stores URLs folowed by `/external-login/google/callback`. For example if you have 2 stores `https://example.com/en` and `https://example.com/fr`, you need to put `https://example.com/en/external-login/google/callback` and `https://example.com/fr/external-login/google/callback` in the `Authorized redirect URIs` field. Also set the `Authorized JavaScript origins` on the same page (in our example above that would be just `https://example.com`).

Reviewer: ```suggestion
You need to configure the allowed redirect URIs under your "OAuth2.0 Client" of your app on the [Google developer console](https://console.developers.google.com/apis/credentials). Fill the `Authorized redirect URIs` field with your stores URLs followed by `/external-login/google/callback`. For example if you have 2 stores `https://example.com/en` and `https://example.com/fr`, you need to put `https://example.com/en/external-login/google/callback` and `https://example.com/fr/external-login/google/callback` in the `Authorized redirect URIs` field. Also set the `Authorized JavaScript origins` on the same page (in our example above that would be just `https://example.com`).
```

(you should consider installing a spell checker to avoid similar mistakes)

### Configuring Authorized domains

You need to add all authorized domains on the [google developer console OAuth Consent Screen](https://console.developers.google.com/apis/credentials/consent/edit). Fill the `Authorized domains` field with your domains (in our example [above](#Configuring-Redirect-URIs) that would be just `example.com`).

### Working with the Google provider in development mode

> Please refer to [Working with the Google provider in development mode](#Working-with-the-Facebook-provider-in-development-mode) above for more details. The only difference is that the environment variable you need to set its `FRONT_COMMERCE_GOOGLE_DEVELOPMENT_BASE_CALLBACK_URL`.

Reviewer 2: ```suggestion
Please refer to [Working with the Facebook provider in development mode](#Working-with-the-Facebook-provider-in-development-mode) above for more details. The only difference is that the environment variable you need to set its `FRONT_COMMERCE_GOOGLE_DEVELOPMENT_BASE_CALLBACK_URL`.
```

> ## Customizing external logins

Reviewer 2: This page is becoming really long, wouldn't it be better to split it (at least for the customization part, maybe even for each provider configuration ?)

Author: yes this is a good idea actually. maybe the customization can go in the advanced section

Author: added and advanced external login page and referred to it. indeed this is much better thx

regarding splitting Facebook and Google in another files I think that wont be necessary after this refactor

Reviewer 2: It is not for now but I was anticipating when we will have more methods

The external logins is made of three parts:
* The login providers. Responsible for handling authenticating the user with the external systems.
* The external login handlers responsible to logging authorized user into our platform (Magento, Proximis, BigCommerce...).
* A module to link the providers and the external login handler.

### Creating your own provider

> Front-Commerce uses passport internally for the Facebook and Google providers though this makes things easier it is neither a requirement nor a recomendation. 

Reviewer 2: ```suggestion
Front-Commerce uses passport internally for the Facebook and Google providers though this makes things easier it is neither a requirement nor a recommendation. 
```

> To create  a provider  you need to register a key and a function with the `auth-external-login` module.

Reviewer 2: ```suggestion
To create a provider you need to register a key and a function with the `auth-external-login` module.
```

```js
import configService from "server/core/config/configService";
import { getCurrentShopConfig } from "server/core/config/getShopConfig";
import { getCurrentShopConfig } from "server/core/config/getShopConfig";
import customConfigProvider from "path_to_your_customConfigProvider";
import { registerExternalLoginProvider } from "auth-external-login/server/providers";

configService.append(customConfigProvider);

const makeCustomProvider = (request, externalLoginHandler, callbackUrl) => {
  const providerConfig = request.config.customProviderConfig;
  if(!providerConfig.clientId || !providerConfig.clientSecret) {
    throw new Error(`Custom provider is not properly configured. clientId and clientSecret are required!`);
  }
  const shopUrl = getCurrentShopConfig(req.config).url; // frequently needed to keep current store selected upon redirection
  return new CustomProvider(externalLoginHandler, providerConfig, {
    shopUrl,
    callbackUrl,
  });
};

registerExternalLoginProvider("custom", makeCustomProvider);
```

> #### The LoginProviderInerface

Reviewer 2: ```suggestion
#### The LoginProviderInterface
```

> To make creating a provider easier, `LoginProviderInerface` can be implemented. By implementing `LoginProviderInerface` you have to implement the 2 aforementioned middleware and you can pick what inputs your constructor takes. You will then register the maker function mentioned above to extract the data needed by your provider and return an instance of it.

Reviewer 2: ```suggestion
To make creating a provider easier, `LoginProviderInterface` can be implemented. By implementing `LoginProviderInterface` you have to implement the 2 aforementioned middleware and you can pick what inputs your constructor takes. You will then register the maker function mentioned above to extract the data needed by your provider and return an instance of it.
```

Reviewer 2: The role of this chapter is unclear to me. Detailling here and after the full detailed reasoning adds too much complexity in my opinion. 

Why not just telling an integrator what to add and its simple role ?
- Handling a new provider and registering it
- Handling a backend that is not implemented and registering it

Author: its our job is to help them and we have these classes/interfaces that could help. Thats why I mention them. Otherwise they can just return an object with the middlewares implemented no problem and all will work fine

Reviewer 2: I think it is better to just tell them to implement the interface then, as it will both enforce the implementation AND send them direct warning while upgrading FC

```
  constructor(externalLoginHandler, providerConfig, { shopUrl, callbackUrl }) {
    ...
  }
```

What is left is implementing `requestLoginMiddleware` which is easily doable using passport:

```
requestLoginMiddleware(req, res, next) {

  const passportAuthenticator = new Passport();

  passportAuthenticator.use(
    new FacebookStrategy(
      {
        clientID: this.providerConfig.clientId,
        clientSecret: this.providerConfig.clientSecret,
        callbackURL: `${this.shopUrl}${this.callbackUrl}`,
        profileFields: ["email", "name"],
      },
      (_, __, profile, cb) => {
        const {
          email,
          first_name: firstname,
          last_name: lastname,
        } = profile._json;

        if (!(email && firstname && lastname)) {
          cb(new MissingInformationError());
        } else {
          cb(null, new ExternalLoginUser(email, firstname, lastname));
        }
      }
    )
  );

  passportAuthenticator.authenticate("facebook", {
    failureRedirect: `${this.shopUrl}/login`,
    session: false,
    scope: ["email"],
  })(req, res, next);
}
```

> ### Creating your own ExternalLoginHandler

Reviewer: not sure anybody but us would want to do this

Author: Proximis/BigCommerce doesnt have these implemented so they would need to build them themselves

Reviewer 2: @mo-fc this is in my opinion the bad reason, it should be our responsibility. 

But this may be handy in some cases using an external user directory.

    // you have access to all your loaders!!! like loaders.Product, loaders.Customer...
  }
}

```

### Putting it all together

As discussed previously the providers provide the middleware to handle the login with the external system, and the external login handler handles actually logging they user in. Now we are going to link the providers with the external login handler in module that we can register into `.front-commerce.js`

A Front-Commerce module is basically a directory that follows certain conventions. It can be placed any where in your folder structure but is commonly placed either under `./modules/your_custom_module_name` or `./src/modules/your_custom_module_name`.

For our purposes we are only interensted in the server part of the module so we need to create `your_custom_module_name/server/module.config.js` file.

The `your_custom_module_name/server/module.config.js` file should export an object with the providers middleware as follows:

```js
import makeExternalLoginEndpoint from "auth-external-login/server/makeExternalLoginEndpoint";
import CustomExternalLoginHandler from "you_external_login_handler/CustomExternalLoginHandler";

export default {

>   endpoint: makeExternalLoginEndpoint((config, staticConfigFromProviders) => new CustomExternalLoginHandler(/** some parameters you extract from the configs */)),

Reviewer: long lines make code snippet hard to read

![image](https://user-images.githubusercontent.com/305563/154314686-9e23bfd7-5b1a-4f63-b0f1-375d992988fa.png)
